### ISSUE TITLE
Migrate Jira SD Issues to GitHub

### ISSUE THREAD

**Issue by dereks** [ocens/hello-world#2]
Just playing with GitHub issue tracker.

**Comment by ocens** [comment]
ok i see that. and i see there are no assignees yet i received an email. presumably because i set up the repository in the first place. for jeff to see this, he would have to be a collaborator and then assigned to this test issue?

**Comment by dereks** [comment]
> This seems onerous unless I’m missing a big piece here?

No, you don't need to invite collaborators to a specific issue. In fact, GitHub has a great feature where you can be typing a comment on an issue, and just use '@' syntax (like @dereks for example) to have me be tagged on that issue and get email notifications from then on. So you will frequently see an issue comment such as

> @kates Did you get this result also?

...and then (a) Kate will get an emailed copy of the comment (by default settings), and (b) she will now get an email on any other comments posted to that issue. No invite needed. That user does not even need to be a contributor ("collaborator") to the project / repository.

Assigning an issue works just like in Jira -- just select the user's name. But I believe to assign an issue, both you and the assignee must be a collaborator of that particular repository. **While anyone can create or comment on an issue, only collaborators can assign issues or be an assignee.**

**Comment by dereks** [comment]
I am probably missing some rules here about private repositories -- I am guessing that, since you have private repos, all of the users (including those commenting on issues) must be invited to your GitHub organization at https://github.com/ocens .

**Comment by dereks** [comment]
I used Kate as an example above, but Kate's actual GitHub account name is

@kwicks

Mine was correct in the example, so I can be tagged with

@dereks 

Currently, this one gets sent to Mark as organization owner:

@ocens 

Once a user has been tagged, or is a collaborator, they will show up in the drop-down selection when you first type the '@' characater.

Another special shortcut is the '#' character, for tagging related issues or pull requests.

**Comment by ocens** [comment]
Yes, that's certainly what i am seeing. i cannot assign someone who is not a collaborator. If, for example, i want to assign a database task to michael i create a repository for the task (easy), define the task (ie issue) that i want him to work on (easy), invite him to be a collaborator on this repository (not difficult-just requires him to accept) and then assign him the issue once he has accepted the invite. he's now responsible for that issue until he assigns it somewhere else. we can comment until the cows come home with the @ prefix with a whole host of people.


Assigning an issue works just like in Jira -- just select the user's name. But I believe to assign an issue, both you and the assignee must be a member of that particular repository. It's been a while, I'm not sure about the rules. This is probably what you are seeing... while anyone can create or comment on an issue, only collaborators can assign issues or be an assignee.

**Comment by ocens** [comment]
@dereks True/False: jeff must signup and pay for a monthly developer account in order to have a github username?

**Comment by ocens** [comment]
Thanks for the assistance with this decision. Let's use the new grib models which i have itemized on Jira by SD#. On Github, since these are all models/work to be added to the grib server, it seems logical to use the gribserver repository on github. As i presently see it, i would add several issues under the gribserver github, with each issue pertaining to one of the tasks presently defined in jIRA?

**Comment by dereks** [comment]
@ocens False, it is not necessary for Jeff to pay for GitHub just to be a developer. GitHub is free for developers to use, unless they want their own private repositories. GitHub is free for hosting public, Open Source repositories. 

But GitHub charges money for an organization to host its own internal (private) repositories. It's $25/month for the first 5 users.

https://github.com/pricing

This is basically the same as paying Atlassian for Jira users.

I don't pay anything to GitHub, but my GitHub account @dereks has been invited to several private repositories owned by my clients, where I do private work for them. If I wanted to host my own internal repos on GitHub (kind of like a 1-person organization), I would need to pay $7/month, or else create an actual organization account, and pay $25 for the first 5 collaborators.

As noted elsewhere, you are currently paying for a 2nd redundant "organization" account (@ocensdev) which we will shut down after migrating any bugs to the final location.

> i would add several issues under the gribserver github

Correct. Your Jira currently has tag names like GRIB_Server, Sidekick, OneMail, ConstantContact. If you  migrate to GitHub, each of those tags would disappear and instead you'd just file the Issue under the appropriate repository (basically, the repo that has the source code that will be changed to fix the issue).

Sometimes there are cross-project issues, in which case we might post an HTTP link to the other project's issue. (Each GitHub Issue gets its own dedicated URL.)

**Comment by ocens** [comment]
OK. Let's do this. I've asked Jeff and Michael to create a GitHub account if they dont already have one. Please proceed to merge the ocens and ocensdev accounts and shut down ocensdev. Also please migrate the SDs under your purview from Jira to GitHub (sidekick, grib_server). I'll then decide what to migrate of that which remains.

**Comment by dereks** [comment]
- Could you make me an "Admin" of both organization accounts?

Here's the HOWTO:

https://stackoverflow.com/questions/21906220/github-how-do-i-make-a-user-an-admin-of-an-organization

I'll need this to transfer ownership of any ocensdev properties to ocens, or to add Jeff and Michael as team members.

I don't want to become an actual (long-term) admin of your GitHub account, so after this is all set up I'd like to be reassigned to a collaborator role.

-  Either Kate or I will copy-paste the entries from Jira, and then flag them as "closed" in Jira (and the issues will still be "open" in GitHub).

Thanks!

**Comment by ocens** [comment]
OK. Give me 30 min or so

**Comment by ocens** [comment]
just sent you an invite to be an 'owner' at ocensdev which appears to be the master account
. i believe thats what you need?

**Comment by ocens** [comment]
ocensdev appears to be the organization. ocens appears as a person. yet it contains multiple seperate repositories. i'll take a look at that now

**Comment by ocens** [comment]
ocens is a personal account. i do not see a way to assign you ownership of that.

**Comment by dereks** [comment]
So it seems that @ocens was set up as a $7/month developer account, including unlimited private repositories.

I think you can keep that account, but I'm not sure. The organization accounts, like @ocensdev, can have teams, sub-teams, but I'm not sure what else. It costs $25/month, for the first 5 users.

Here is a list of the differences. It is a business decision over which account you want to use for your company:

https://help.github.com/articles/differences-between-user-and-organization-accounts/

**Comment by dereks** [comment]
I just saw there are many repos under @ocensdev. I think GitHub is worth $25/month. Therefor, I slightly  recommend keeping @ocensdev, and then assign all of the private repos currently owned by @ocens to the organization @ocensdev. Then your organization will own it all, and @ocens will just be another user.

Please note, I will need at least one robot account (for use in deployment scripts) that has access to the repos and can do automated checkouts and deployments. That account will be a free GitHub account, but I don't know offhand if it will need to be one of the 5 collaborator accounts you get for a $25/month as an organization.

**Comment by ocens** [comment]
$7 is less than $25. We've never used teams since we started it. if its not something you see much in your experience then i see no reason to keep paying for ocensdev organization. bigger question is can you merge from ocensdev into ocens?

**Comment by ocens** [comment]
and as i was typing your alternate rec came in. that's fine. $25 is not much more than $7. let's get it squared away and end this confusion. keep ocensdev

**Comment by dereks** [comment]
Agree, part of that $25/month is corporate-level support if some kind of problem comes up. It's worth it.

Next, decide if @ocens will become your own personal GitHub account (so ocens = Mark, in particular), or if you want to create a new GitHub account so that @ocens can always forward to "The Current GitHub Administrator of OCENS".

Or, @ocens could become the new robot account I need for automated scripts.

**Comment by ocens** [comment]
alright. just transferred one of less important ocens repositories to ocensdev as a test. i guess i'll need to do the rest one at a time

**Comment by ocens** [comment]
let @ocens be the robot but we should change the email address assigned to it so i can use my freeberg account with my personal dev. i'll do that and alias it to my freeberg account

**Comment by dereks** [comment]
Sounds good. Note that the robot account (@ocens or any other) will also need an email Inbox (or even better: simple email forwarding address) of its own. 

I need this so I can do things like register new SSH keys on GitHub for the robot, to be used in automated deployments. GitHub will demand a valid working email address.

For now, the robot's email address could just forward to dereks@cool-st.com . But once the deployments are all completed, it should forward to you or Jeff, in case GitHub emails you something important (like "Notice, your account has been hacked/closed/suspended", or "We require you to reset your password", etc.). When Microsoft bought GitHub they invalidated all my SSH keys, for example.

**Comment by ocens** [comment]
i have moved all but 3 repositories (the hello-world where this thread is located) and two others which appear empty.

**Comment by dereks** [comment]
To confirm that the repos are truly empty, make sure there is only one branch in the repo: *master*, which is the default branch.

Sometimes there is a repo where the default branch (*master*) is empty, but a different branch has important stuff. You should see a dropdown called "Branch:", with only one option, "master".

Also make sure there are zero issues for the project. Maybe it was just an empty repo being used solely for the issue tracker function of GitHub.

If the repo is truly empty (only one branch, called *master*, and no files, no issues) then it can safely be deleted.

**Comment by dereks** [comment]
ocensgit test email has appeared in my Inbox.

![ocensdev-email](https://user-images.githubusercontent.com/1031825/46982235-57b2be00-d090-11e8-923c-30cc30877f8e.png)

**Comment by ocens** [comment]
ocensgit@ocens.com is the primary address now for the @ocens account. it forwards to dereks@cool-st.com

**Comment by dereks** [comment]
Awaiting GitHub user names for Mark, Michael, and Jeff...

**Comment by ocens** [comment]
jeffdotthomassen
cozmotion

**Comment by dereks** [comment]
Kate moved over all GRIB_Server and Sidekick issues from Jira to their respective GitHub issue trackers. **All transferred tickets have been flagged as Done in Jira.**

There were some that she could not identify, which might still be GRIB_Server tickets. These remain Open in Jira, waiting for Mark to transfer them (or to ask me/Kate to transfer them). I will leave this ticket open until I have confirmation that all GRIB_Server and Sidekick tickets have been transferred.

Kate and I will start working the Sidekick and GRIB_Server tickets in GitHub. In a couple of hours I will create a new repo for the server deployment scripts, and log some issues for myself to resolve.

**Comment by ocens** [comment]
Yes, that’s how I encountered the repository question. I was starting to work through those to clarify which of the Sidekick are high on the list and which can wait until after the model work is done. Was returned a 404 error for every one of them.

Mark

Mark Freeberg
OCENS
Saving Lives, Saving Time and Supporting Sustainable Adventure

Satellite Systems And Service
Weather, Email, Voice & Data Solutions
Inmarsat, Iridium, Globalstar, KVH, VSAT
___________________________________________________________________
Office: 206.878.8270

**Comment by harleyfrey** [comment]
I would have thought if i was invited to comment on an issue i would not then encounter a 404 error when i attempt to do so

**Comment by dereks** [comment]
Yes, the 404 is a confusing security thing GitHub does. Instead of telling any attacking scannerbots, "Yes, there's good stuff here, you're just not allowed to see it", it tells the bots "What? I have no idea what you're talking about. 404 Not Found!" so they don't gain any information by scanning.

Your new account @harleyfrey has not been a part of any GitHub tickets yet (you were using @ocens yesterday) and I have not yet added you to the OCENS organization. I sent you a standalone email confirming some users...

**Comment by harleyfrey** [comment]
Just a bit frightful when you expect to see 20 to 30 repositories and find just 1 when all those repositories are the life of your business! Glad you and Kate were involved

**Comment by dereks** [comment]
Update: All Members have been added. @harleyfrey is an owner. @dereks is still listed as an owner temporarily, expecting to move to "Outside Collaborator" by the end of the month.
